File: java-driver/src/test/java/github/hacimertgokhan/drivers/FakeServer.java
package github.hacimertgokhan.drivers;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiFunction;

/**
 * In-process stand-in for a Denis server speaking protocol 2 in JSON mode.
 * One thread per connection; replies are written in order, one line each.
 * Tests can intercept lines with {@link #override}.
 */
final class FakeServer implements AutoCloseable {
    static final String GROUP = "grp";
    static final String PASSWORD = "pass with spaces";

    final ServerSocket server;
    final List<Conn> connections = new CopyOnWriteArrayList<>();
    /** Every received line of every connection, in global arrival order. */
    final List<String> arrivals = new CopyOnWriteArrayList<>();
    final Map<String, String> data = new ConcurrentHashMap<>();
    final Map<String, Long> ttl = new ConcurrentHashMap<>();
    final AtomicInteger projects = new AtomicInteger();
    final CountDownLatch released = new CountDownLatch(1);
    /** Return a reply line (or "" for no reply) to intercept a command; null falls through to the default handler. */
    volatile BiFunction<Conn, String, String> override;
    /** When set, accepted connections are closed immediately. */
    volatile boolean refuse;
    private final Thread acceptor;
    private volatile boolean closed;

    FakeServer() throws IOException {
        server = new ServerSocket(0, 50, InetAddress.getLoopbackAddress());
        acceptor = new Thread(this::acceptLoop, "fake-denis-acceptor");
        acceptor.setDaemon(true);
        acceptor.start();
    }

    int port() {
        return server.getLocalPort();
    }

    DenisClient.Builder builder() {
        return DenisClient.builder().host("127.0.0.1").port(port());
    }

    Conn connection(int index) {
        return connections.get(index);
    }

    /** Commands received on all connections, in arrival order per connection. */
    List<String> allReceived() {
        List<String> all = new ArrayList<>();
        for (Conn c : connections) {
            all.addAll(c.received);
        }
        return all;
    }

    private void acceptLoop() {
        while (!closed) {
            try {
                Socket s = server.accept();
                if (refuse) {
                    s.close();
                    continue;
                }
                Conn c = new Conn(s, connections.size());
                connections.add(c);
                Thread t = new Thread(c::serve, "fake-denis-conn-" + c.index);
                t.setDaemon(true);
                t.start();
            } catch (IOException e) {
                return;
            }
        }
    }

    @Override
    public void close() throws IOException {
        closed = true;
        released.countDown();
        server.close();
        for (Conn c : connections) {
            c.close();
        }
    }

    static String ok(String fields) {
        return fields.isEmpty() ? "{\"ok\":true}" : "{\"ok\":true," + fields + "}";
    }

    static String error(String code, String message) {
        return "{\"ok\":false,\"error\":" + Json.quote(message) + ",\"code\":\"" + code + "\"}";
    }

    final class Conn {
        final Socket socket;
        final int index;
        final List<String> received = new CopyOnWriteArrayList<>();
        private OutputStream out;
        boolean json;
        boolean loggedIn;
        String group;
        String token;

        Conn(Socket socket, int index) {
            this.socket = socket;
            this.index = index;
        }

        void serve() {
            try (Socket s = socket) {
                BufferedReader in = new BufferedReader(new InputStreamReader(s.getInputStream(), StandardCharsets.UTF_8));
                out = s.getOutputStream();
                String line;
                while ((line = in.readLine()) != null) {
                    received.add(line);
                    arrivals.add(line);
                    String reply = null;
                    BiFunction<Conn, String, String> o = override;
                    if (o != null) {
                        reply = o.apply(this, line);
                    }
                    if (reply == null) {
                        reply = handle(line);
                    }
                    if (socket.isClosed()) {
                        return;
                    }
                    if (!reply.isEmpty()) {
                        send(reply + "\n");
                    }
                }
            } catch (IOException ignored) {
                // client went away
            }
        }

        synchronized void send(String raw) throws IOException {
            out.write(raw.getBytes(StandardCharsets.UTF_8));
            out.flush();
        }

        void close() {
            try {
                socket.close();
            } catch (IOException ignored) {
                // closing
            }
        }

        /** Block this connection (like a stalled server) until the test ends. */
        String hang() {
            try {
                released.await(30, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            return "";
        }

        private String handle(String line) {
            String trimmed = line.stripLeading();
            String[] w = trimmed.split(" ", 3);
            String cmd = w[0].toUpperCase();
            switch (cmd) {
                case "MODE":
                    json = true;
                    return ok("\"message\":\"mode json\"");
                case "PING":
                    return ok("\"message\":\"PONG\"");
                case "HELLO":
                    return ok("\"server\":\"denis\",\"version\":\"0.1.0\",\"protocol\":2,\"features\":[\"json\",\"sql-params\"],\"loggedIn\":" + loggedIn);
                case "LIN": {
                    String[] p = trimmed.split(" ", 3);
                    if (p.length == 3 && p[1].equals(GROUP) && p[2].equals(PASSWORD)) {
                        loggedIn = true;
                        group = p[1];
                        return ok("\"message\":\"Logged in to group: grp\",\"group\":\"grp\",\"admin\":true");
                    }
                    return error("AUTH", "Login failed: unknown group or wrong password");
                }
                default:
                    break;
            }
            if (!loggedIn) {
                return error("NOAUTH", "Please login first using LIN command");
            }
            switch (cmd) {
                case "AUTH":
                    if (w[1].equalsIgnoreCase("CREATE")) {
                        return ok("\"message\":\"Project created\",\"token\":\"tok-" + projects.incrementAndGet() + "\"");
                    }
                    if (w[1].equalsIgnoreCase("DELETE")) {
                        if (w[2].equals(token)) {
                            token = null;
                        }
                        return ok("\"message\":\"Project deleted\"");
                    }
                    if (w[1].startsWith("bad")) {
                        return error("AUTH", "Cannot auth with: " + w[1]);
                    }
                    token = w[1];
                    return ok("\"message\":\"Authenticated to project: " + token + "\"");
                case "WHOAMI":
                    return ok("\"group\":\"grp\",\"admin\":true,\"project\":" + (token == null ? "null" : Json.quote(token)));
                case "INFO":
                    return ok("\"info\":{\"server\":{\"version\":\"0.1.0\"},\"stats\":{\"commands\":5}}");
                case "PROJECTS":
                    return ok("\"projects\":[{\"token\":\"tok-1\",\"owner\":\"grp\",\"keys\":3,\"tables\":1,\"current\":true},"
                            + "{\"token\":\"legacy\",\"owner\":null,\"keys\":0,\"tables\":0,\"current\":false}],\"count\":2");
                case "SAVE":
                    return ok("\"message\":\"Snapshot written\",\"bytes\":10,\"records\":2,\"millis\":1");
                case "BACKUP":
                    return ok("\"message\":\"Backup created\",\"name\":\"b1.zip\",\"path\":\"/tmp/b1.zip\",\"bytes\":99,\"createdAt\":\"2026-09-23T10:00:00Z\"");
                case "BACKUPS":
                    return ok("\"backups\":[{\"name\":\"b1.zip\",\"path\":\"/tmp/b1.zip\",\"bytes\":99,\"createdAt\":\"2026-09-23T10:00:00Z\"}],\"directory\":\"/tmp\"");
                default:
                    break;
            }
            if (token == null) {
                return error("NOPROJECT", "Please authenticate first using AUTH command");
            }
            switch (cmd) {
                case "GET": {
                    String v = data.get(w[1]);
                    if (v == null) {
                        return "{\"ok\":false,\"key\":" + Json.quote(w[1]) + ",\"error\":\"not found\",\"code\":\"NOTFOUND\"}";
                    }
                    return ok("\"key\":" + Json.quote(w[1]) + ",\"data\":" + Json.quote(v));
                }
                case "SET":
                case "UPDATE": {
                    // exactly like the real server: flags are removed, everything else is kept verbatim
                    String value = w[2];
                    if (value.contains("-&")) {
                        StringBuilder sb = new StringBuilder();
                        boolean first = true;
                        for (String word : w[2].stripTrailing().split(" ", -1)) {
                            if (!word.startsWith("-&")) {
                                sb.append(first ? "" : " ").append(word);
                                first = false;
                            }
                        }
                        value = sb.toString();
                    }
                    data.put(w[1], value);
                    return ok("\"message\":\"Ok (Cache)\"");
                }
                case "DEL":
                    return ok("\"message\":\"Ok\",\"deleted\":" + (data.remove(w[1]) != null));
                case "EXISTS":
                    return ok("\"key\":" + Json.quote(w[1]) + ",\"exists\":" + data.containsKey(w[1]) + ",\"cache\":true,\"persistent\":false");
                case "MGET": {
                    Map<String, Object> m = new LinkedHashMap<>();
                    for (String k : trimmed.substring(5).split(" ")) {
                        m.put(k, data.get(k));
                    }
                    return ok("\"data\":" + Json.write(m));
                }
                case "INCR":
                case "DECR": {
                    String[] p = trimmed.split(" ");
                    long delta = p.length > 2 && !p[2].startsWith("-&") ? Long.parseLong(p[2]) : 1;
                    long v = Long.parseLong(data.getOrDefault(p[1], "0")) + (cmd.equals("DECR") ? -delta : delta);
                    data.put(p[1], Long.toString(v));
                    return ok("\"key\":" + Json.quote(p[1]) + ",\"data\":\"" + v + "\",\"value\":" + v);
                }
                case "EXPIRE":
                    ttl.put(w[1], (long) (Double.parseDouble(w[2]) * 1000));
                    return ok("\"key\":" + Json.quote(w[1]) + ",\"data\":\"1\",\"updated\":" + data.containsKey(w[1]));
                case "PERSIST":
                    ttl.remove(w[1]);
                    return ok("\"key\":" + Json.quote(w[1]) + ",\"data\":\"1\",\"updated\":" + data.containsKey(w[1]));
                case "TTL": {
                    long ms = !data.containsKey(w[1]) ? -2 : ttl.getOrDefault(w[1], -1L);
                    long s = ms < 0 ? ms : (ms + 999) / 1000;
                    return ok("\"key\":" + Json.quote(w[1]) + ",\"data\":\"" + s + "\",\"ttl\":" + s + ",\"ttlMillis\":" + ms);
                }
                case "DBSIZE":
                    return ok("\"keys\":" + data.size() + ",\"cache\":" + data.size() + ",\"persistent\":0,\"tables\":0,\"data\":\"" + data.size() + "\"");
                case "HEAVEN":
                    data.clear();
                    return ok("\"message\":\"Ok.\"");
                case "KEYS":
                    return ok("\"keys\":" + Json.write(new ArrayList<>(new java.util.TreeSet<>(data.keySet()))) + ",\"count\":" + data.size() + ",\"truncated\":false");
                case "QUERY": {
                    Map<String, Object> q = Json.parseObject(trimmed.substring(6));
                    String sql = (String) q.get("sql");
                    if (sql.startsWith("SELECT")) {
                        return ok("\"columns\":[\"sql\",\"params\"],\"rows\":[[" + Json.quote(sql) + "," + Json.quote(Json.write(q.get("params")))
                                + "]],\"count\":1,\"data\":\"[]\"");
                    }
                    if (sql.startsWith("BAD")) {
                        return "{\"ok\":false,\"error\":\"Table not found: t\",\"code\":\"SQL\",\"data\":\"ERROR: Table not found: t\"}";
                    }
                    return ok("\"message\":\"1 row inserted\",\"affected\":1,\"lastRowId\":7,\"data\":\"OK: 1 row inserted\"");
                }
                case "DUMP":
                    return ok("\"format\":1,\"cache\":" + Json.write(new java.util.TreeMap<>(data)) + ",\"persistent\":{},\"ttl\":{},\"tables\":{}");
                case "IMPORT": {
                    Map<String, Object> d = Json.parseObject(trimmed.substring(7));
                    long cache = d.get("cache") instanceof Map ? ((Map<?, ?>) d.get("cache")).size() : 0;
                    long persistent = d.get("persistent") instanceof Map ? ((Map<?, ?>) d.get("persistent")).size() : 0;
                    long tables = d.get("tables") instanceof Map ? ((Map<?, ?>) d.get("tables")).size() : 0;
                    long rows = 0;
                    if (d.get("tables") instanceof Map) {
                        for (Object def : ((Map<?, ?>) d.get("tables")).values()) {
                            Object r = def instanceof Map ? ((Map<?, ?>) def).get("rows") : null;
                            rows += r instanceof List ? ((List<?>) r).size() : 0;
                        }
                    }
                    return ok("\"message\":\"Imported\",\"imported\":{\"persistent\":" + persistent + ",\"cache\":" + cache
                            + ",\"tables\":" + tables + ",\"rows\":" + rows + "}");
                }
                default:
                    return error("UNKNOWN", "Unknown command: " + cmd);
            }
        }
    }
}

File: clients/node/test/fake-server.js
"use strict";

// An in-process fake Denis server for the unit tests (no real server needed).

const net = require("node:net");

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

const MAIN_TOKEN = "main-token";

/**
 * A fake server speaking enough of the JSON-mode protocol (docs/PROTOCOL.md):
 * MODE, LIN, AUTH (CREATE/DELETE), PING, HELLO, HELP, INFO, WHOAMI,
 * GET/SET/UPDATE/DEL/EXISTS/KEYS/MGET/INCR/HEAVEN, IMPORT, a tiny SQL
 * (CREATE TABLE, INSERT with bound params, SELECT, SHOW TABLES, DESCRIBE,
 * DROP TABLE) through `SQL` and `QUERY {"sql","params"}`, the GraphQL-shaped
 * `QUERY { ... }` (echoed back), ADMIN <main-token> ..., plus test commands:
 * ECHO <text>, SPLIT <text> (reply written byte by byte), DELAY <ms> <text>,
 * SLOW (never answers), KILL (drops the socket), FAIL <code>.
 * Lines of one connection are handled strictly one after another, like Denis.
 */
async function fakeServer(options = {}) {
  const users = options.users || { ci: "pass word" };
  const state = {
    projects: new Map([["tok0", new Map()], ["tok1", new Map()]]),
    tables: new Map(), // token -> Map(name -> {columns: [{name, type}], rows: [object]})
    quotas: new Map(),
    created: 0,
    conns: [],
    imports: [],
    refuse: options.refuse || false,
  };

  function tablesOf(token) {
    if (!state.tables.has(token)) state.tables.set(token, new Map());
    return state.tables.get(token);
  }

  function sqlError(message) {
    return { ok: false, error: message, code: "SQL", data: `ERROR: ${message}` };
  }

  function tableInfo(name, table) {
    return { name, columns: table.columns, rows: table.rows.length };
  }

  /** A very small SQL: enough to check reply shapes and wire lines. */
  function runSql(token, sql, params = []) {
    const tables = tablesOf(token);
    const text = sql.replace(/\s+/g, " ").trim().replace(/;$/, "");
    let m;
    if ((m = /^CREATE TABLE (?:IF NOT EXISTS )?(\w+) \((.*)\)$/i.exec(text))) {
      if (tables.has(m[1])) {
        if (/IF NOT EXISTS/i.test(text)) return { ok: true, type: "affected", affected: 0, message: "Table exists", lastRowId: 0, data: "OK: Table exists" };
        return sqlError(`Table already exists: ${m[1]}`);
      }
      const columns = m[2].split(",").map((c) => {
        const [name, type = "TEXT"] = c.trim().split(" ");
        return { name, type: type.toUpperCase() };
      });
      tables.set(m[1], { columns, rows: [] });
      return { ok: true, type: "affected", affected: 0, message: "Table created", lastRowId: 0, data: "OK: Table created" };
    }
    if ((m = /^INSERT INTO (\w+) \(([^)]*)\) VALUES \(([^)]*)\)$/i.exec(text))) {
      const table = tables.get(m[1]);
      if (!table) return sqlError(`Table not found: ${m[1]}`);
      const cols = m[2].split(",").map((c) => c.trim());
      let p = 0;
      const values = m[3].split(",").map((v) => {
        v = v.trim();
        if (v === "?") return params[p++];
        if (/^'.*'$/.test(v)) return v.slice(1, -1);
        return Number(v);
      });
      const row = {};
      for (const c of table.columns) row[c.name] = null;
      cols.forEach((c, i) => (row[c] = values[i]));
      table.rows.push(row);
      return { ok: true, type: "affected", affected: 1, message: "1 row inserted", lastRowId: table.rows.length, data: "OK: 1 row inserted" };
    }
    if ((m = /^SELECT (.+) FROM (\w+)(?: WHERE (\w+) = (\?|\d+|'[^']*'))?$/i.exec(text))) {
      const table = tables.get(m[2]);
      if (!table) return sqlError(`Table not found: ${m[2]}`);
      const columns = m[1] === "*" ? table.columns.map((c) => c.name) : m[1].split(",").map((c) => c.trim());
      let rows = table.rows;
      if (m[3]) {
        const want = m[4] === "?" ? params[0] : /^'/.test(m[4]) ? m[4].slice(1, -1) : Number(m[4]);
        rows = rows.filter((r) => r[m[3]] === want);
      }
      const out = rows.map((r) => Object.fromEntries(columns.map((c) => [c, r[c] ?? null])));
      return { ok: true, type: "rows", columns, rows: out, count: out.length, data: JSON.stringify(out) };
    }
    if (/^SHOW TABLES$/i.test(text)) {
      const list = [...tables].map(([name, t]) => tableInfo(name, t));
      return { ok: true, type: "tables", tables: list, count: list.length };
    }
    if ((m = /^(?:DESCRIBE|DESC) (\w+)$/i.exec(text))) {
      const table = tables.get(m[1]);
      if (!table) return sqlError(`Table not found: ${m[1]}`);
      return { ok: true, type: "tables", tables: [tableInfo(m[1], table)], count: 1 };
    }
    if ((m = /^DROP TABLE (\w+)$/i.exec(text))) {
      if (!tables.delete(m[1])) return sqlError(`Table not found: ${m[1]}`);
      return { ok: true, type: "affected", affected: 0, message: "Table dropped", lastRowId: 0, data: "OK: Table dropped" };
    }
    return sqlError("Unsupported SQL query");
  }

  function usageOf(token) {
    const store = state.projects.get(token);
    const quota = state.quotas.get(token) || { maxKeys: 0, maxBytes: 0 };
    return {
      token,
      usage: { cachedKeys: store.size, cachedBytes: 0, persistedKeys: 0, persistedBytes: 0 },
      quota,
    };
  }

  function admin(words) {
    if (words[0] !== MAIN_TOKEN) return { ok: false, error: "ADMIN refused: wrong main token", code: "AUTH" };
    const [, sub, token, a, b] = words;
    switch ((sub || "").toUpperCase()) {
      case "LIST":
        return { ok: true, projects: [...state.projects.keys()].map(usageOf), count: state.projects.size };
      case "CREATE": {
        const created = `adm${++state.created}`;
        state.projects.set(created, new Map());
        if (token !== undefined) state.quotas.set(created, { maxKeys: Number(token), maxBytes: Number(a) });
        return { ok: true, message: "Project created", token: created };
      }
      case "IMPORT": {
        const added = !state.projects.has(token);
        if (added) state.projects.set(token, new Map());
        if (a !== undefined) state.quotas.set(token, { maxKeys: Number(a), maxBytes: Number(b) });
        return { ok: true, message: added ? "Project imported" : "Project already known", token, added };
      }
      case "USAGE":
        if (!state.projects.has(token)) return { ok: false, error: `Unknown project: ${token}` };
        return { ok: true, ...usageOf(token) };
      case "QUOTA":
        if (!state.projects.has(token)) return { ok: false, error: `Unknown project: ${token}` };
        state.quotas.set(token, { maxKeys: Number(a), maxBytes: Number(b) });
        return { ok: true, message: "Quota set", ...usageOf(token) };
      case "FLUSH":
        state.projects.get(token).clear();
        return { ok: true, message: "Project flushed" };
      case "DROP":
        state.projects.delete(token);
        return { ok: true, message: "Project dropped" };
      default:
        return { ok: false, error: "USAGE: ADMIN <main-token> LIST|CREATE|IMPORT|USAGE|QUOTA|FLUSH|DROP", code: "USAGE" };
    }
  }

  function handle(conn, line) {
    const space = line.indexOf(" ");
    const cmd = (space < 0 ? line : line.slice(0, space)).toUpperCase();
    const args = space < 0 ? "" : line.slice(space + 1);
    const words = args.split(" ");
    switch (cmd) {
      case "MODE":
        return { ok: true, message: "mode json" };
      case "PING":
        return { ok: true, message: "PONG" };
      case "HELLO":
        return { ok: true, server: "denis", version: "1.1.0-fake", protocol: 2, features: ["json", "sql-params", "graph"], loggedIn: conn.loggedIn };
      case "HELP":
        return {
          ok: true,
          commands: [
            { name: "PING", usage: "PING", description: "Liveness check", needsLogin: false, needsProject: false },
            { name: "SQL", usage: "SQL <statement>", description: "Run SQL", needsLogin: true, needsProject: true },
          ],
        };
      case "EXIT":
        return () => conn.socket.end(JSON.stringify({ ok: true, message: "Bye." }) + "\n");
      case "ECHO":
        return { ok: true, data: args };
      case "SPLIT":
        return async () => {
          const bytes = Buffer.from(JSON.stringify({ ok: true, data: args }) + "\n");
          for (const byte of bytes) {
            conn.socket.write(Buffer.from([byte]));
            await sleep(1);
          }
        };
      case "DELAY":
        return sleep(Number(words[0])).then(() => ({ ok: true, data: words.slice(1).join(" ") }));
      case "SLOW":
        return new Promise(() => {});
      case "KILL":
        return () => conn.socket.destroy();
      case "FAIL":
        return { ok: false, error: `failed with ${words[0]}`, code: words[0] };
      case "ADMIN":
        return admin(words);
      case "LIN": {
        const group = words[0];
        const password = args.slice(group.length + 1);
        if (users[group] !== undefined && users[group] === password) {
          conn.loggedIn = true;
          return { ok: true, message: `Logged in to group: ${group}`, group, admin: true };
        }
        return { ok: false, error: "Login failed: unknown group or wrong password", code: "AUTH" };
      }
      default:
        break;
    }
    if (!conn.loggedIn) return { ok: false, error: "Please login first using LIN command", code: "NOAUTH" };
    if (cmd === "AUTH") {
      if (words[0] === "CREATE") {
        const token = `new${++state.created}`;
        state.projects.set(token, new Map());
        return { ok: true, message: "Project created", token };
      }
      if (words[0] === "DELETE") {
        state.projects.delete(words[1]);
        if (conn.token === words[1]) conn.token = null;
        return { ok: true, message: "Project deleted" };
      }
      if (!state.projects.has(words[0])) return { ok: false, error: `Cannot auth with: ${words[0]}`, code: "AUTH" };
      conn.token = words[0];
      return { ok: true, message: `Authenticated to project: ${words[0]}` };
    }
    if (cmd === "WHOAMI") return { ok: true, group: "ci", admin: true, project: conn.token };
    if (cmd === "INFO") {
      const project = conn.token && state.projects.has(conn.token) ? { ...usageOf(conn.token).usage, quota: usageOf(conn.token).quota } : undefined;
      return {
        ok: true,
        version: "1.1.0-fake",
        uptimeSeconds: 5,
        startedAt: "2026-01-01T00:00:00Z",
        connections: { open: state.conns.length, total: state.conns.length },
        commandsTotal: 10,
        cacheKeys: 0,
        persistedKeys: 0,
        projects: state.projects.size,
        group: "ci",
        ...(project ? { project } : {}),
        memory: { usedMb: 1, maxMb: 2 },
        info: { server: { version: "1.1.0-fake", protocol: 2 }, stats: { commands: 10 } },
      };
    }
    const store = state.projects.get(conn.token);
    if (!store) return { ok: false, error: "Please authenticate first using AUTH command", code: "NOPROJECT" };
    const key = words[0];
    switch (cmd) {
      case "SET":
      case "UPDATE": {
        const rest = args.slice(key.length + 1);
        const value = rest.split(" ").filter((w) => !w.startsWith("-&")).join(" ");
        const quota = state.quotas.get(conn.token);
        if (quota && quota.maxKeys && !store.has(key) && store.size >= quota.maxKeys) {
          return { ok: false, error: `quota exceeded: keys (limit ${quota.maxKeys})`, code: "QUOTA", resource: "keys", limit: quota.maxKeys };
        }
        store.set(key, value);
        return { ok: true, message: "Ok (Cache)" };
      }
      case "GET":
        if (key === "__oom") return { ok: false, error: "out of memory", code: "OOM" };
        return store.has(key) ? { ok: true, key, data: store.get(key) } : { ok: false, key, error: "not found", code: "NOTFOUND" };
      case "DEL": {
        const existed = store.delete(key);
        return { ok: true, message: "Ok (Cache,Protobuf).", deleted: existed };
      }
      case "EXISTS":
        return { ok: true, key, exists: store.has(key), cache: store.has(key), persistent: false };
      case "KEYS": {
        const re = new RegExp(`^${key.replace(/[.+^${}()|\\]/g, "\\$&").replace(/\*/g, ".*").replace(/\?/g, ".")}$`);
        const keys = [...store.keys()].filter((k) => re.test(k)).sort();
        return { ok: true, keys, count: keys.length, truncated: false };
      }
      case "MGET": {
        const values = Object.fromEntries(words.map((k) => [k, store.has(k) ? store.get(k) : null]));
        return { ok: true, data: values, values };
      }
      case "HEAVEN": {
        const removed = store.size;
        store.clear();
        return { ok: true, message: "Ok.", removed };
      }
      case "INCR": {
        const value = Number(store.get(key) || 0) + (words[1] && !words[1].startsWith("-&") ? Number(words[1]) : 1);
        store.set(key, String(value));
        return { ok: true, key, data: String(value), value };
      }
      case "SAVE":
        return { ok: true, message: "Snapshot written", bytes: 1, records: 1, millis: 1 };
      case "SQL":
        return runSql(conn.token, args);
      case "QUERY": {
        let bound = null;
        if (args.startsWith("{\"")) {
          try {
            bound = JSON.parse(args);
          } catch {
            bound = null;
          }
        }
        if (bound && typeof bound.sql === "string") return runSql(conn.token, bound.sql, bound.params || []);
        if (!args.startsWith("{")) return { ok: false, error: "syntax: expected {", offset: 0 };
        return { ok: true, data: { document: args }, errors: [{ path: "bad", error: "unknown resolver" }] };
      }
      case "IMPORT": {
        const data = JSON.parse(args);
        state.imports.push({ line, data });
        const tables = Object.values(data.tables || {});
        return {
          ok: true,
          message: "Imported",
          imported: {
            persistent: Object.keys(data.persistent || {}).length,
            cache: Object.keys(data.cache || {}).length,
            tables: tables.length,
            rows: tables.reduce((n, t) => n + (t.rows || []).length, 0),
          },
        };
      }
      default:
        return { ok: false, error: `Unknown command: ${cmd} (try HELP)`, code: "UNKNOWN" };
    }
  }

  const sockets = new Set();
  const server = net.createServer((socket) => {
    sockets.add(socket);
    socket.on("close", () => sockets.delete(socket));
    socket.on("error", () => {});
    socket.setNoDelay(true);
    if (state.refuse) {
      socket.end(JSON.stringify({ ok: false, error: "max number of clients reached", code: "LIMIT" }) + "\n");
      return;
    }
    const conn = { socket, lines: [], loggedIn: false, token: null, outstanding: 0, maxOutstanding: 0 };
    state.conns.push(conn);
    let buffer = Buffer.alloc(0);
    let chain = Promise.resolve();
    socket.on("data", (chunk) => {
      buffer = Buffer.concat([buffer, chunk]);
      let index;
      while ((index = buffer.indexOf(10)) >= 0) {
        const line = buffer.subarray(0, index).toString("utf8").replace(/\r$/, "").trim();
        buffer = buffer.subarray(index + 1);
        if (!line) continue;
        conn.lines.push(line);
        conn.outstanding++;
        conn.maxOutstanding = Math.max(conn.maxOutstanding, conn.outstanding);
        chain = chain
          .then(() => handle(conn, line))
          .then(async (reply) => {
            conn.outstanding--;
            if (socket.destroyed) return;
            if (typeof reply === "function") await reply();
            else socket.write(JSON.stringify(reply) + "\n");
          });
      }
    });
  });
  await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
  state.port = server.address().port;
  state.server = server;
  /** Every line received on any connection. */
  state.allLines = () => state.conns.flatMap((c) => c.lines);
  /** Drop every client connection (the server keeps listening). */
  state.kill = () => {
    for (const socket of sockets) socket.destroy();
  };
  /** Stop listening and drop every connection. */
  state.close = () =>
    new Promise((resolve) => {
      state.kill();
      server.close(() => resolve());
    });
  return state;
}

module.exports = { fakeServer, sleep, MAIN_TOKEN };
